molgenis-data/src/main/java/org/molgenis/data/listeners/EntityListenersService.java

```diff
@@ -0,0 +1,140 @@
+package org.molgenis.data.listeners;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
+import org.molgenis.data.Entity;
+import org.molgenis.data.MolgenisDataException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.PreDestroy;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+
+@Service
+public final class EntityListenersService
+{
+	private final Logger LOG = LoggerFactory.getLogger(EntityListenersService.class);
+	private final ConcurrentHashMap<String, SetMultimap<Object, EntityListener>> entityListeners = new ConcurrentHashMap<>();
+
+	/**
+	 * Register a repository to the entity listeners service once
+	 *
+	 * @param repoFullName
+	 * @return
+	 */
+	void register(String repoFullName)
+	{
+		if (!entityListeners.containsKey(requireNonNull(repoFullName)))
+		{
+			entityListeners.put(repoFullName, HashMultimap.create());
+		}
+	}
+
+	/**
+	 * Update all registered listeners of the entities
+	 *
+	 * @param repoFullName
+	 * @param entities
+	 * @return Stream<Entity>
+	 */
+	Stream<Entity> updateEntities(String repoFullName, Stream<Entity> entities)
+	{
+		verifyRepoRegistered(repoFullName);
+		SetMultimap<Object, EntityListener> entityListeners = this.entityListeners.get(repoFullName);
+		return entities.filter(entity -> {
+			Set<EntityListener> entityEntityListeners = entityListeners.get(entity.getIdValue());
+			entityEntityListeners.forEach(entityListener -> {
+				entityListener.postUpdate(entity);
```

Comment:
This one reads through the listeners without locking it (which is good cause otherwise you cannot do concurrent updates) but now I *think* it may still happen that the addListener() method is concurrently adding entries to the map while you iterate through it with this forEach.

In practice I suspect the add() will only be called at startup so we won't run into this bug.